blogposts/2020/find-and-replace.md

```diff
@@ -0,0 +1,582 @@
+
+Here are a few more examples:
+
+| Description | Comby invocation |
+|-|-|
+| Reverse function argument order | `comby 'myFunc(:[1], :[2])' 'myFunc(:[2], :[1])'` |
+| Convert HTML table to JSON | `comby '<tr><td>:[1]</td><td>:[2]</td></tr>' '":[1]": :[2]'` |
+| Update Go error handling to use wrapped errors | `comby 'fmt.Errorf(":[head]%s:[tail]", err)' 'fmt.Errorf(":[head]%w:[tail]", err)'` |
+
+
+Comby's advantages can be summed up in two words: expressivity and ergonomics.
+* Expressivity: You can capture patterns that involve nested delimiters, which are inexpressible with regex.
+* Ergonomics: Unlike regex, Comby has very few special characters, so Comby patterns are generally
+  more readable and writable.
+
+Due to its newness, Comby is not yet as widely adopted as regex-based tools like `grep` and `sed`,
+but it is off to a strong start and I predict it will supplant regex in more and more use cases in
+the coming years.
+
+
+
+
+## Beyond your local machine
+
+Find-and-replace over files on your local machine is one thing, but there are scenarios when you'd
+like to apply a transformation across a universe of code that exists beyond your local
+filesystem. If you're working in enterprise, this might be something like adopting a consistent
+logging standard across all services in your application. If you're working in open source, this
+might be updating all the callers of a now-deprecated library function that has hundreds of
+downstream dependencies. In either case, you may find that the trivial becomes intractable at a
+large enough scale.
+
+From many Sourcegraph customers, we've heard that even small changes to a shared API can take months
+or even years to fully propagate across the entire codebase of an organization. This is a big enough
+problem that many dev organizations have created tools to facilitate large-scale changes to code
+that span many files, teams, and repositories. These tools and the associated processes go by
+various names such as "large-scale refactoring", "large-scale codemods", and "code
+shepherding". These tools are usually specific to the organization that created them and rarely
+released publicly.
+
+### Sourcegraph Campaigns
+
+After hearing about many similar challenges from many of our customers, our team at Sourcegraph
+decided to productize a solution that could generally handle the problem of large-scale code
+transformation.
+
+We're calling this feature Sourcegraph Campaigns. If Sourcegraph Code Search is the "find" operation
+over all your code, then Sourcegraph Campaigns is the "replace".
+
+Suppose your organization has decided it wants to standardize error handling across your
+codebase. In particular, you want to use [wrapped errors](https://blog.golang.org/go1.13-errors) in
+Go code. This means you want to ensure that instances of `fmt.Errorf` use the `%w` format verb,
+rather than `%s`.
+
+Without Campaigns, this would involve cloning down all your repositories one-by-one, running a
```

Comment:
lowercase campaigns, don't refer to things as "Sourcegraph xyz" above. It's ok and actually helpful to do that in marketing copy, but not in this kind of post IMO.